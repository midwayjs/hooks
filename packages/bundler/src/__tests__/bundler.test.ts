import {
  All,
  Api,
  ApiRouter,
  BaseTrigger,
  Delete,
  Get,
  Head,
  Operator,
  OperatorType,
  Options,
  parseApiModule,
  Patch,
  Post,
  Put,
} from '@midwayjs/hooks-core'
import { AbstractBundlerAdapter } from '..'
import { wrap } from 'jest-snapshot-serializer-raw'

const router = new ApiRouter()

class TestBundlerAdapter extends AbstractBundlerAdapter {
  constructor() {
    super({
      name: 'test',
      router,
    })
  }

  getSource(): string {
    return '/'
  }
}

it('create bundler', async () => {
  const testBundlerAdapter = new TestBundlerAdapter()
  expect(testBundlerAdapter).toBeTruthy()
  expect(testBundlerAdapter).toBeInstanceOf(AbstractBundlerAdapter)
  expect(testBundlerAdapter.getName()).toEqual('test')
})

it('generate client', async () => {
  const testBundlerAdapter = new TestBundlerAdapter()
  expect(testBundlerAdapter.generateClient([])).toMatchSnapshot()

  const apis = parseApiModule(
    {
      get: Api(Get(), async () => {}),
      post: Api(Post(), async () => {}),
      put: Api(Put(), async () => {}),
      del: Api(Delete(), async () => {}),
      patch: Api(Patch(), async () => {}),
      head: Api(Head(), async () => {}),
      options: Api(Options(), async () => {}),
      default: Api(All(), async () => {}),
    },
    '/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(wrap(testBundlerAdapter.generateClient(apis))).toMatchSnapshot()
})

it('with requestClient should not generate client', async () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
        })
      },
    }
  }

  const testBundlerAdapter = new TestBundlerAdapter()

  const apis = parseApiModule(
    {
      custom: Api(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(testBundlerAdapter.generateClient(apis)).toMatchSnapshot()
})

it('with requestClient generate client', async () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty<BaseTrigger>(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
          requestClient: {
            fetcher: 'ws',
            client: '@midwayjs/rpc',
          },
        })
      },
    }
  }

  const testBundlerAdapter = new TestBundlerAdapter()

  const apis = parseApiModule(
    {
      get: Api(Get(), async () => {}),
      custom: Api(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(testBundlerAdapter.generateClient(apis)).toMatchSnapshot()
})

it('with requestClient generate multi client', async () => {
  const CustomTrigger = (): Operator<void> => {
    return {
      name: 'Custom',
      metadata({ setMetadata: defineProperty }) {
        defineProperty<BaseTrigger>(OperatorType.Trigger, {
          type: 'Custom',
          isCustom: true,
          requestClient: {
            fetcher: 'cloud',
            client: 'cloud-invoker',
          },
        })
      },
    }
  }

  const testBundlerAdapter = new TestBundlerAdapter()

  const apis = parseApiModule(
    {
      get: Api(Get(), async () => {}),
      custom: Api(CustomTrigger(), async () => {}),
    },
    '/server/api/index.ts',
    testBundlerAdapter.getRouter()
  )

  expect(wrap(testBundlerAdapter.generateClient(apis))).toMatchSnapshot()
})
