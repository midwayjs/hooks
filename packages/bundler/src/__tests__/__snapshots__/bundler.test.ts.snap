// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate client 1`] = `"// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten."`;

exports[`generate client 2`] = `
// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.
import { http as rpc$0 } from '@midwayjs/rpc'

        export  function get (...args) {
          const route = get.route;
          return rpc$0(...args, route)
        };
        get.route = {
          trigger: {"type":"HTTP","method":"GET","path":"/get"},
          functionId: 'index-get',
          useInputMetadata: false,
        };
      

        export  function post (...args) {
          const route = post.route;
          return rpc$0(...args, route)
        };
        post.route = {
          trigger: {"type":"HTTP","method":"POST","path":"/post"},
          functionId: 'index-post',
          useInputMetadata: false,
        };
      

        export  function put (...args) {
          const route = put.route;
          return rpc$0(...args, route)
        };
        put.route = {
          trigger: {"type":"HTTP","method":"PUT","path":"/put"},
          functionId: 'index-put',
          useInputMetadata: false,
        };
      

        export  function del (...args) {
          const route = del.route;
          return rpc$0(...args, route)
        };
        del.route = {
          trigger: {"type":"HTTP","method":"DELETE","path":"/del"},
          functionId: 'index-del',
          useInputMetadata: false,
        };
      

        export  function patch (...args) {
          const route = patch.route;
          return rpc$0(...args, route)
        };
        patch.route = {
          trigger: {"type":"HTTP","method":"PATCH","path":"/patch"},
          functionId: 'index-patch',
          useInputMetadata: false,
        };
      

        export  function head (...args) {
          const route = head.route;
          return rpc$0(...args, route)
        };
        head.route = {
          trigger: {"type":"HTTP","method":"HEAD","path":"/head"},
          functionId: 'index-head',
          useInputMetadata: false,
        };
      

        export  function options (...args) {
          const route = options.route;
          return rpc$0(...args, route)
        };
        options.route = {
          trigger: {"type":"HTTP","method":"OPTIONS","path":"/options"},
          functionId: 'index-options',
          useInputMetadata: false,
        };
      

        export default function $default (...args) {
          const route = $default.route;
          return rpc$0(...args, route)
        };
        $default.route = {
          trigger: {"type":"HTTP","method":"ALL","path":"/"},
          functionId: 'index',
          useInputMetadata: false,
        };
      
`;

exports[`with requestClient generate client 1`] = `
"// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.
import { http as rpc$0 } from '@midwayjs/rpc'
import { ws as rpc$1 } from '@midwayjs/rpc'

        export  function get (...args) {
          const route = get.route;
          return rpc$0(...args, route)
        };
        get.route = {
          trigger: {\\"type\\":\\"HTTP\\",\\"method\\":\\"GET\\",\\"path\\":\\"/server/api/get\\"},
          functionId: 'server-api-index-get',
          useInputMetadata: false,
        };
      

        export  function custom (...args) {
          const route = custom.route;
          return rpc$1(...args, route)
        };
        custom.route = {
          trigger: {\\"type\\":\\"Custom\\",\\"isCustom\\":true},
          functionId: 'server-api-index-custom',
          useInputMetadata: false,
        };
      "
`;

exports[`with requestClient generate multi client 1`] = `
// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten.
import { http as rpc$0 } from '@midwayjs/rpc'
import { cloud as rpc$1 } from 'cloud-invoker'

        export  function get (...args) {
          const route = get.route;
          return rpc$0(...args, route)
        };
        get.route = {
          trigger: {"type":"HTTP","method":"GET","path":"/server/api/get"},
          functionId: 'server-api-index-get',
          useInputMetadata: false,
        };
      

        export  function custom (...args) {
          const route = custom.route;
          return rpc$1(...args, route)
        };
        custom.route = {
          trigger: {"type":"Custom","isCustom":true},
          functionId: 'server-api-index-custom',
          useInputMetadata: false,
        };
      
`;

exports[`with requestClient should not generate client 1`] = `"// This file is auto-generated by @midwayjs/bundler. Any changes will be overwritten."`;
